### PR TITLE
Updates for Fastify v5

### PR DESCRIPTION
**Pull request by mcollina**
I had to add a bunch of `@ts-expect-error` to make things pass, as I have no time to understand why the types here are not behaving.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by bencoder** [comment]
Typing issues are due to conflicting differences between `@fastify/session` and `@fastify/secure-session` (which does not have `regenerate` method).

There is a TODO in the code here that would go some way to fixing it: https://github.com/fastify/fastify-passport/blob/3a05ee0f33f1fe26bca22bbea7880382202ebea7/src/session-managers/SecureSessionManager.ts#L57-L58

Have raised [issue in fastify-secure-session](https://github.com/fastify/fastify-secure-session/issues/233) to action this TODO so we can tidy that up.

If there's any urgency to this PR then we could approve it as-is with a follow-up ticket here to clean up once the types are improved on secure-session

**Comment by robwalkerco** [comment]
I've also had a brief investigation of the types issues, and have created a patch that resolves them by more explicitly typing the `request` data as `FastifyRequest & { session: FastifySessionObject | Session<any> }`, and then using some type guarding to ensure that the necessary methods exist before calling them etc.

The following patch resolves the type issues, while continuing to pass all the tests.

```patch
diff --git forkSrcPrefix/src/session-managers/SecureSessionManager.ts forkDstPrefix/src/session-managers/SecureSessionManager.ts
index 1bb468ddd80459a1b9b0cbdc750c6ee042b60694..8a87c36c4376d6cbf33e2f0090d1a7135b12052b 100644
--- forkSrcPrefix/src/session-managers/SecureSessionManager.ts
+++ forkDstPrefix/src/session-managers/SecureSessionManager.ts
@@ -1,8 +1,11 @@
 /* eslint-disable @typescript-eslint/no-unsafe-argument */
-/// <reference types="@fastify/secure-session" />
 import { FastifyRequest } from 'fastify'
 import { AuthenticateOptions } from '../AuthenticationRoute'
 import { SerializeFunction } from '../Authenticator'
+import { FastifySessionObject } from '@fastify/session'
+import { Session } from '@fastify/secure-session'
+
+type Request = FastifyRequest & { session: FastifySessionObject | Session<any> }
 
 /** Class for storing passport data in the session using `@fastify/secure-session` or `@fastify/session` */
 export class SecureSessionManager {
@@ -35,21 +38,18 @@ export class SecureSessionManager {
     }
   }
 
-  async logIn(request: FastifyRequest, user: any, options?: AuthenticateOptions) {
+  async logIn(request: Request, user: any, options?: AuthenticateOptions) {
     const object = await this.serializeUser(user, request)
 
     // Handle @fastify/session to prevent token/CSRF fixation
-    // @ts-expect-error no idea why this does not get typed correctly
     if (request.session.regenerate) {
       if (this.clearSessionOnLogin && object) {
         const keepSessionInfoKeys: string[] = [...this.clearSessionIgnoreFields]
         if (options?.keepSessionInfo) {
           keepSessionInfoKeys.push(...Object.keys(request.session))
         }
-        // @ts-expect-error no idea why this does not get typed correctly
         await request.session.regenerate(keepSessionInfoKeys)
       } else {
-        // @ts-expect-error no idea why this does not get typed correctly
         await request.session.regenerate()
       }
     }
@@ -57,31 +57,26 @@ export class SecureSessionManager {
     // TODO: This is quite hacky. The best option would be having a regenerate method
     // on secure-session as well
     else if (this.clearSessionOnLogin && object) {
-      const currentFields = request.session.data() || {}
+      const currentData = 'data' in request.session ? request.session.data() : undefined
+      const currentFields = currentData ?? {}
       for (const field of Object.keys(currentFields)) {
         if (options?.keepSessionInfo || this.clearSessionIgnoreFields.includes(field)) {
           continue
         }
-        // @ts-expect-error we don't know the field
         request.session.set(field, undefined)
       }
     }
-    // eslint-disable @typescript-eslint/no-unsafe-argument
-    // @ts-expect-error we don't know the field
     request.session.set(this.key, object)
   }
 
-  async logOut(request: FastifyRequest) {
-    // @ts-expect-error we don't know the field
+  async logOut(request: Request) {
     request.session.set(this.key, undefined)
-    // @ts-expect-error no idea why this does not get typed correctly
     if (request.session.regenerate) {
-      // @ts-expect-error no idea why this does not get typed correctly
       await request.session.regenerate()
     }
   }
 
-  getUserFromSession(request: FastifyRequest) {
+  getUserFromSession(request: Request) {
     return request.session.get(this.key)
   }
 }
diff --git forkSrcPrefix/src/type-extensions.ts forkDstPrefix/src/type-extensions.ts
index 5835aca6aff580d9cd83a42b7df21e66861bdd82..8cc25cb9433619f766145315a603d07ac3d14a1a 100644
--- forkSrcPrefix/src/type-extensions.ts
+++ forkDstPrefix/src/type-extensions.ts
@@ -19,6 +19,10 @@ declare module 'fastify' {
    */
   interface PassportUser {}
 
+  interface ExpressSessionData {
+    [key: string]: any
+  }
+
   interface FastifyRequest {
     flash: ReturnType<typeof flashFactory>['request']
 
diff --git forkSrcPrefix/src/AuthenticationRoute.ts forkDstPrefix/src/AuthenticationRoute.ts
index 79e109f451709e8d05a0e3205cd3986a494032f7..945cceea17f013a9910da87481b356c4a0f90449 100644
--- forkSrcPrefix/src/AuthenticationRoute.ts
+++ forkDstPrefix/src/AuthenticationRoute.ts
@@ -1,4 +1,3 @@
-/// <reference types="@fastify/secure-session" />
 import * as http from 'http'
 import AuthenticationError from './errors'
 import Authenticator from './Authenticator'
```

**Comment by mcollina** [comment]
@dancastillo @Uzlopak @climba03003 fancy a review?

File: src/session-managers/SecureSessionManager.ts [hunk 1]
```diff
@@ -53,7 +56,8 @@ export class SecureSessionManager {
     // TODO: This is quite hacky. The best option would be having a regenerate method
     // on secure-session as well
     else if (this.clearSessionOnLogin && object) {
-      const currentFields = request.session.data() || {}
+      const currentData: SessionData = request.session?.data() ?? {}
+      const currentFields = currentData ?? {}
       for (const field of Object.keys(currentFields)) {
```
Comment (dancastillo) [inline src/session-managers/SecureSessionManager.ts:61]: I think we can remove initalizing `currentFields` and just pass `currentData` to `Object.keys(currentData)`
```suggestion
      const currentData: SessionData = request.session?.data() ?? {}
      for (const field of Object.keys(currentData)) {
```

**Review by dancastillo — APPROVED** [review]
LGTM - just one small comment

**Pull request merged**